src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2824,56 +2913,146 @@ private static bool IsReallyAType(TypeSymbol type)
             {
-                return;
+                get { return _dictionary.Count; }
+            }
+
+            internal void AddAll(HashSet<TypeSymbolWithAnnotations> set, ConversionsBase conversions)
+            {
+                foreach (var candidate in set)
+                {
+                    Add(candidate, conversions);
+                }
+            }
+
+            internal void Add(TypeSymbolWithAnnotations type, ConversionsBase conversions)
+            {
+                if (Merge(type, conversions))
+                {
+                    return;
+                }
+
+                List<TypeSymbolWithAnnotations> list;
+                if (!_dictionary.TryGetValue(type, out list))
+                {
+                    list = new List<TypeSymbolWithAnnotations>();
+                    _dictionary.Add(type, list);
+                }
+                list.Add(type);
+            }
+
+            // Returns true if type was found and merged.
+            internal bool Merge(TypeSymbolWithAnnotations type, ConversionsBase conversions)
+            {
+                // We make an exception when type is dynamic, for backwards compatibility 
+                if (type.IsDynamic())
+                {
+                    return false;
+                }
+
+                List<TypeSymbolWithAnnotations> list;
+                if (!_dictionary.TryGetValue(type, out list))
+                {
+                    return false;
+                }
+
+                bool found = list.Contains(type, _valueComparer);
+                for (int i = 0; i < list.Count; i++)
+                {
+                    list[i] = MergeTupleNames(MergeDynamic(list[i], type, conversions.CorLibrary), type);
+                }
+                return found;
             }
 
-            if (dict.TryGetValue(old, out TypeSymbol latest))
+            internal void Remove(TypeSymbolWithAnnotations type)
             {
-                dict.Remove(old);
-                TypeSymbol merged = MergeTupleNames(MergeDynamic(latest, @new, _conversions.CorLibrary), @new);
-                dict.Add(merged, merged);
+                List<TypeSymbolWithAnnotations> list;
+                if (!_dictionary.TryGetValue(type, out list))
+                {
+                    return;
+                }
+
+                var builder = new List<TypeSymbolWithAnnotations>();
+                foreach (var value in list)
+                {
+                    if (!_valueComparer.Equals(type, value))
+                    {
+                        builder.Add(value);
+                    }
+                }
+                if (builder.Count == 0)
+                {
+                    _dictionary.Remove(type);
+                }
+                else
+                {
+                    _dictionary[type] = builder;
+                }
             }
         }
 
+        private static TypeCompareKind GetComparison(bool includeNullability)
```

Comment:
I'm not sure this method is worth it. Only used in two places below (once with includeNullability `true` and once `false`).

Reply:
It's also used to compare candidates in the last loop in `Fix`.